WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -139,24 +142,33 @@ private List<AliceClient> CreateAliceClients(RoundState roundState)
 				var secret = hdKey.PrivateKey.GetBitcoinSecret(Keymanager.GetNetwork());
 				aliceClients.Add(new AliceClient(roundState.Id, aliceArenaClient, coin, roundState.FeeRate, secret));
 			}
-			return aliceClients;
+			return aliceClients.ToImmutable();
 		}
 
-		private async Task RegisterCoinsAsync(IEnumerable<AliceClient> aliceClients, CancellationToken cancellationToken)
+		private async Task<ImmutableArray<AliceClient>> RegisterCoinsAsync(ImmutableArray<AliceClient> aliceClients, CancellationToken cancellationToken)
 		{
-			async Task RegisterInputTask(AliceClient aliceClient)
+			var successfulAlices = ImmutableArray.CreateBuilder<AliceClient>();
+
+			foreach (var aliceClient in aliceClients.OrderByDescending(x => x.Coin.Amount))
 			{
-				await aliceClient.RegisterInputAsync(cancellationToken).ConfigureAwait(false);
+				try
+				{
+					await aliceClient.RegisterInputAsync(cancellationToken).ConfigureAwait(false);
+					successfulAlices.Add(aliceClient);
+				}
+				catch (Exception e)
```

Comment:
I think we will need to work with `SmartCoin`s instead of simply `Coin`s here because depending on the type of error we will need to update the `SmartCoin` status accordingly. For example, if the coin was registered okay then we have to mar it as `CoinjoinInProgress` (or something like that, I don't remember the property name right now), if it fails with an `AlreadySpent` with have to mark it as `AlreadySpentAccordingToBackend` and so on.

Reply:
Good idea, later we can change it.